File: arkanoid3d.go
// arkanoid3d.go — Арканоид 3D на Go

package main

import (
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"time"

	"github.com/eiannone/keyboard"
)

const W = 10
const H = 8
const D = 5
const PADDLE_W = 5

var blocks [][][]int
var paddleX, paddleZ int
var ballX, ballY, ballZ int
var ballDX, ballDY, ballDZ int
var score, lives, level, highScore int
var gameOver, paused, ballLaunched bool
var randSrc *rand.Rand
var lastUpdate time.Time

func clear() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func loadHighScore() { highScore = 0 }
func saveHighScore() {}

func generateBlocks() {
	blocks = make([][][]int, W)
	for x := 0; x < W; x++ {
		blocks[x] = make([][]int, H)
		for y := 0; y < H; y++ {
			blocks[x][y] = make([]int, D)
			for z := 0; z < D; z++ {
				if y < H-1 {
					blocks[x][y][z] = ((x+y+z)%2 == 0) ? randSrc.Intn(2)+1 : 0
				}
			}
		}
	}
}

func resetBall() {
	ballX = paddleX
	ballY = H - 2
	ballZ = paddleZ
	ballDX = 1
	ballDY = -1
	ballDZ = 0
	ballLaunched = false
}

func launchBall() {
	if !ballLaunched {
		ballLaunched = true
	}
}

func movePaddle(dx, dz int) {
	nx := paddleX + dx
	nz := paddleZ + dz
	if nx >= 0 && nx < W && nz >= 0 && nz < D {
		paddleX = nx
		paddleZ = nz
		if !ballLaunched {
			ballX = paddleX
			ballZ = paddleZ
		}
	}
}

func update() {
	if gameOver || paused || !ballLaunched {
		return
	}
	ballX += ballDX
	ballY += ballDY
	ballZ += ballDZ

	if ballX <= 0 || ballX >= W-1 {
		ballDX *= -1
		ballX += ballDX * 2
	}
	if ballZ <= 0 || ballZ >= D-1 {
		ballDZ *= -1
		ballZ += ballDZ * 2
	}
	if ballY <= 0 {
		ballDY *= -1
		ballY += ballDY * 2
	}

	if ballY >= H-1 {
		if ballX >= paddleX-PADDLE_W/2 && ballX <= paddleX+PADDLE_W/2 && ballZ == paddleZ {
			ballDY *= -1
			ballY = H - 2
			dx := ballX - paddleX
			if dx == 0 {
				ballDX = 1
			} else {
				ballDX = dx
			}
			ballDZ = randSrc.Intn(3) - 1
		} else {
			lives--
			if lives <= 0 {
				gameOver = true
				if score > highScore {
					highScore = score
				}
			} else {
				resetBall()
			}
		}
	}

	if ballX >= 0 && ballX < W && ballY >= 0 && ballY < H && ballZ >= 0 && ballZ < D {
		if blocks[ballX][ballY][ballZ] > 0 {
			blocks[ballX][ballY][ballZ]--
			if blocks[ballX][ballY][ballZ] == 0 {
				score += 10 * level
			}
			// отражение
			if ballX > 0 && blocks[ballX-1][ballY][ballZ] > 0 {
				ballDX *= -1
			} else if ballX < W-1 && blocks[ballX+1][ballY][ballZ] > 0 {
				ballDX *= -1
			} else if ballY > 0 && blocks[ballX][ballY-1][ballZ] > 0 {
				ballDY *= -1
			} else if ballY < H-1 && blocks[ballX][ballY+1][ballZ] > 0 {
				ballDY *= -1
			} else if ballZ > 0 && blocks[ballX][ballY][ballZ-1] > 0 {
				ballDZ *= -1
			} else if ballZ < D-1 && blocks[ballX][ballY][ballZ+1] > 0 {
				ballDZ *= -1
			} else {
				ballDY *= -1
			}
			ballX += ballDX
			ballY += ballDY
			ballZ += ballDZ
		}
	}

	blocksLeft := false
	for x := 0; x < W; x++ {
		for y := 0; y < H-1; y++ {
			for z := 0; z < D; z++ {
				if blocks[x][y][z] > 0 {
					blocksLeft = true
				}
			}
		}
	}
	if !blocksLeft {
		level++
		generateBlocks()
		resetBall()
	}
}

func draw() {
	clear()
	fmt.Printf("🧱 ARKANOID 3D  |  Счёт: %d  |  Уровень: %d  |  Жизни: %d  |  Рекорд: %d\n", score, level, lives, highScore)
	if paused {
		fmt.Println("⏸ ПАУЗА")
	}
	for z := D - 1; z >= 0; z-- {
		fmt.Printf("\nСлой %d (z=%d):\n", z, z)
		fmt.Print("+" + strings.Repeat("-", W*2) + "+\n")
		for y := 0; y < H; y++ {
			fmt.Print("|")
			for x := 0; x < W; x++ {
				if blocks[x][y][z] > 0 {
					fmt.Print("██")
				} else if x == ballX && y == ballY && z == ballZ {
					fmt.Print("● ")
				} else {
					fmt.Print("  ")
				}
			}
			fmt.Println("|")
		}
		fmt.Print("+" + strings.Repeat("-", W*2) + "+\n")
	}
	fmt.Println("\nПлатформа (вид сверху):")
	fmt.Print("  ")
	for x := 0; x < W; x++ {
		if x >= paddleX-PADDLE_W/2 && x <= paddleX+PADDLE_W/2 {
			fmt.Print("=")
		} else {
			fmt.Print(" ")
		}
	}
	fmt.Println("\nУправление: WASD - движение, Пробел - запуск/пауза, Q - выход")
}

func main() {
	randSrc = rand.New(rand.NewSource(time.Now().UnixNano()))
	loadHighScore()
	generateBlocks()
	paddleX = W / 2
	paddleZ = D / 2
	resetBall()
	score = 0
	lives = 3
	level = 1
	gameOver = false
	paused = false

	if err := keyboard.Open(); err != nil {
		panic(err)
	}
	defer keyboard.Close()

	lastUpdate = time.Now()
	for !gameOver {
		draw()
		_, key, err := keyboard.GetKey()
		if err != nil {
			continue
		}
		if key == keyboard.KeyEsc || key == 'q' || key == 'Q' {
			break
		}
		if key == 'p' || key == 'P' {
			paused = !paused
			continue
		}
		if !paused {
			switch key {
			case 'w', 'W':
				movePaddle(0, -1)
			case 's', 'S':
				movePaddle(0, 1)
			case 'a', 'A':
				movePaddle(-1, 0)
			case 'd', 'D':
				movePaddle(1, 0)
			case ' ':
				if !ballLaunched {
					launchBall()
				} else {
					paused = !paused
				}
			}
			if time.Since(lastUpdate).Seconds() > 0.05 {
				update()
				lastUpdate = time.Now()
			}
		}
		time.Sleep(20 * time.Millisecond)
	}
	fmt.Printf("ИГРА ОКОНЧЕНА! Счёт: %d\n", score)
	saveHighScore()
}

File: arkanoid3d.rs
// arkanoid3d.rs — Арканоид 3D на Rust

use std::io::{self, Write, stdout};
use std::thread;
use std::time::{Duration, Instant};
use rand::Rng;
use termion::{clear, cursor, color, style};
use termion::input::TermRead;

const W: usize = 10;
const H: usize = 8;
const D: usize = 5;
const PADDLE_W: i32 = 5;

struct Game {
    blocks: Vec<Vec<Vec<i32>>>,
    paddle_x: i32,
    paddle_z: i32,
    ball_x: i32,
    ball_y: i32,
    ball_z: i32,
    ball_dx: i32,
    ball_dy: i32,
    ball_dz: i32,
    score: i32,
    lives: i32,
    level: i32,
    high_score: i32,
    game_over: bool,
    paused: bool,
    ball_launched: bool,
}

impl Game {
    fn new() -> Self {
        let mut game = Game {
            blocks: vec![vec![vec![0; D]; H]; W],
            paddle_x: (W/2) as i32,
            paddle_z: (D/2) as i32,
            ball_x: (W/2) as i32,
            ball_y: (H-2) as i32,
            ball_z: (D/2) as i32,
            ball_dx: 1,
            ball_dy: -1,
            ball_dz: 0,
            score: 0,
            lives: 3,
            level: 1,
            high_score: 0,
            game_over: false,
            paused: false,
            ball_launched: false,
        };
        game.load_high_score();
        game.generate_blocks();
        game
    }

    fn load_high_score(&mut self) {
        self.high_score = 0;
    }

    fn generate_blocks(&mut self) {
        let mut rng = rand::thread_rng();
        for x in 0..W {
            for y in 0..H-1 {
                for z in 0..D {
                    self.blocks[x][y][z] = if (x+y+z) % 2 == 0 { rng.gen_range(1..=2) } else { 0 };
                }
            }
        }
    }

    fn reset_ball(&mut self) {
        self.ball_x = self.paddle_x;
        self.ball_y = (H-2) as i32;
        self.ball_z = self.paddle_z;
        self.ball_dx = 1;
        self.ball_dy = -1;
        self.ball_dz = 0;
        self.ball_launched = false;
    }

    fn launch_ball(&mut self) {
        if !self.ball_launched {
            self.ball_launched = true;
        }
    }

    fn move_paddle(&mut self, dx: i32, dz: i32) {
        let nx = self.paddle_x + dx;
        let nz = self.paddle_z + dz;
        if nx >= 0 && nx < W as i32 && nz >= 0 && nz < D as i32 {
            self.paddle_x = nx;
            self.paddle_z = nz;
            if !self.ball_launched {
                self.ball_x = self.paddle_x;
                self.ball_z = self.paddle_z;
            }
        }
    }

    fn update(&mut self) {
        if self.game_over || self.paused || !self.ball_launched { return; }
        self.ball_x += self.ball_dx;
        self.ball_y += self.ball_dy;
        self.ball_z += self.ball_dz;

        if self.ball_x <= 0 || self.ball_x >= W as i32 -1 {
            self.ball_dx *= -1;
            self.ball_x += self.ball_dx * 2;
        }
        if self.ball_z <= 0 || self.ball_z >= D as i32 -1 {
            self.ball_dz *= -1;
            self.ball_z += self.ball_dz * 2;
        }
        if self.ball_y <= 0 {
            self.ball_dy *= -1;
            self.ball_y += self.ball_dy * 2;
        }

        if self.ball_y >= H as i32 -1 {
            let half = PADDLE_W / 2;
            if self.ball_x >= self.paddle_x - half && self.ball_x <= self.paddle_x + half && self.ball_z == self.paddle_z {
                self.ball_dy *= -1;
                self.ball_y = H as i32 - 2;
                let dx = self.ball_x - self.paddle_x;
                self.ball_dx = if dx == 0 { 1 } else { dx };
                let mut rng = rand::thread_rng();
                self.ball_dz = rng.gen_range(-1..=1);
            } else {
                self.lives -= 1;
                if self.lives <= 0 {
                    self.game_over = true;
                    if self.score > self.high_score { self.high_score = self.score; }
                } else {
                    self.reset_ball();
                }
            }
        }

        if self.ball_x >= 0 && self.ball_x < W as i32 && self.ball_y >= 0 && self.ball_y < H as i32 && self.ball_z >= 0 && self.ball_z < D as i32 {
            let bx = self.ball_x as usize;
            let by = self.ball_y as usize;
            let bz = self.ball_z as usize;
            if self.blocks[bx][by][bz] > 0 {
                self.blocks[bx][by][bz] -= 1;
                if self.blocks[bx][by][bz] == 0 {
                    self.score += 10 * self.level;
                }
                // отражение
                if bx > 0 && self.blocks[bx-1][by][bz] > 0 {
                    self.ball_dx *= -1;
                } else if bx < W-1 && self.blocks[bx+1][by][bz] > 0 {
                    self.ball_dx *= -1;
                } else if by > 0 && self.blocks[bx][by-1][bz] > 0 {
                    self.ball_dy *= -1;
                } else if by < H-1 && self.blocks[bx][by+1][bz] > 0 {
                    self.ball_dy *= -1;
                } else if bz > 0 && self.blocks[bx][by][bz-1] > 0 {
                    self.ball_dz *= -1;
                } else if bz < D-1 && self.blocks[bx][by][bz+1] > 0 {
                    self.ball_dz *= -1;
                } else {
                    self.ball_dy *= -1;
                }
                self.ball_x += self.ball_dx;
                self.ball_y += self.ball_dy;
                self.ball_z += self.ball_dz;
            }
        }

        let mut blocks_left = false;
        for x in 0..W {
            for y in 0..H-1 {
                for z in 0..D {
                    if self.blocks[x][y][z] > 0 { blocks_left = true; }
                }
            }
        }
        if !blocks_left {
            self.level += 1;
            self.generate_blocks();
            self.reset_ball();
        }
    }

    fn draw(&self) {
        print!("{}{}", clear::All, cursor::Goto(1,1));
        println!("🧱 ARKANOID 3D  |  Счёт: {}  |  Уровень: {}  |  Жизни: {}  |  Рекорд: {}", self.score, self.level, self.lives, self.high_score);
        if self.paused { println!("⏸ ПАУЗА"); }
        for z in (0..D).rev() {
            println!("\nСлой {} (z={}):", z, z);
            print!("+");
            for _ in 0..W*2 { print!("-"); }
            println!("+");
            for y in 0..H {
                print!("|");
                for x in 0..W {
                    if self.blocks[x][y][z] > 0 {
                        print!("██");
                    } else if x as i32 == self.ball_x && y as i32 == self.ball_y && z as i32 == self.ball_z {
                        print!("● ");
                    } else {
                        print!("  ");
                    }
                }
                println!("|");
            }
            print!("+");
            for _ in 0..W*2 { print!("-"); }
            println!("+");
        }
        println!("\nПлатформа (вид сверху):");
        print!("  ");
        for x in 0..W {
            let x_i = x as i32;
            if x_i >= self.paddle_x - PADDLE_W/2 && x_i <= self.paddle_x + PADDLE_W/2 && self.paddle_z == z as i32 {
                print!("=");
            } else {
                print!(" ");
            }
        }
        println!("\nУправление: WASD - движение, Пробел - запуск/пауза, Q - выход");
        stdout().flush().unwrap();
    }

    fn run(&mut self) {
        let stdin = io::stdin();
        let mut keys = stdin.keys();
        let mut last_update = Instant::now();
        while !self.game_over {
            self.draw();
            if let Some(Ok(key)) = keys.next() {
                match key {
                    termion::event::Key::Char('q') => break,
                    termion::event::Key::Char('p') => self.paused = !self.paused,
                    termion::event::Key::Char('w') => self.move_paddle(0, -1),
                    termion::event::Key::Char('s') => self.move_paddle(0, 1),
                    termion::event::Key::Char('a') => self.move_paddle(-1, 0),
                    termion::event::Key::Char('d') => self.move_paddle(1, 0),
                    termion::event::Key::Char(' ') => {
                        if !self.ball_launched {
                            self.launch_ball();
                        } else {
                            self.paused = !self.paused;
                        }
                    }
                    _ => {}
                }
            }
            if last_update.elapsed().as_secs_f64() > 0.05 {
                self.update();
                last_update = Instant::now();
            }
            thread::sleep(Duration::from_millis(20));
        }
        println!("ИГРА ОКОНЧЕНА! Счёт: {}", self.score);
    }
}

fn main() {
    let mut game = Game::new();
    game.run();
}
